rust/src/flatfit/mod.rs

```diff
@@ -0,0 +1,129 @@
+use crate::FifoWindow;
+use alga::general::AbstractMonoid;
+use alga::general::Operator;
+use std::cell::RefCell;
+use std::marker::PhantomData;
+
+const LOW_CAP: usize = 2;
+
+#[derive(Clone)]
+pub struct FlatFIT<Value, BinOp>
+where
+    Value: AbstractMonoid<BinOp> + Clone,
+    BinOp: Operator,
+{
+    front: usize,
+    back: usize,
+    size: usize,
+    buffer: RefCell<Vec<Item<Value>>>,
+    tracing_indices: RefCell<Vec<usize>>,
+    binop: PhantomData<BinOp>,
+}
+
+#[derive(Clone)]
+struct Item<Value> {
+    val: Value,
+    next: usize,
+}
+
+impl<Value> Item<Value> {
+    fn new(val: Value, next: usize) -> Self {
+        Self { val, next }
+    }
+}
+
+impl<Value, BinOp> FifoWindow<Value, BinOp> for FlatFIT<Value, BinOp>
+where
+    Value: AbstractMonoid<BinOp> + Clone,
+    BinOp: Operator,
+{
+    fn new() -> Self {
+        Self {
+            front: 0,
+            back: 0,
+            size: 0,
+            buffer: RefCell::new(Vec::new()),
+            tracing_indices: RefCell::new(Vec::new()),
+            binop: PhantomData,
+        }
+    }
+    fn push(&mut self, val: Value) {
+        let capacity = self.buffer.borrow().capacity();
+        if self.size + 1 > capacity {
+            self.rescale(capacity * 2)
+        }
+        let mut buffer = self.buffer.borrow_mut();
+        self.size += 1;
+        let prev = self.back;
+        self.back = (self.back + 1) % self.size;
+        buffer[self.back].val = val;
+        buffer[prev].next = self.back;
```

Comment:
Is this correct when we push into an empty window? In that case, we start out with `front == back == size == 0`. When we end, `back == 1`, `size == 1`, `buffer[1]` contains `val` but `front == 0`. If we then do a query, doesn't that mean when we'll look at the contents of `buffer[0].val`, which is still uninitialized?

In our implementation, when the window has only one element, then `front == back`. That's what starting with `back == -1` accomplishes, although it also requires the awkwardness of indexing using a signed integer (which is not allowed in Rust, for good reason).

Our implementation has a guard around assigning the next pointer at the end of inserting/pushing for the case of adding an element to an empty window. Maybe your code needs a guard around incrementing `back`? That is, if `back == front && size == 0`, then don't increment `back`.

Reply:
Hmm, I might be wrong, I thought the checks for when `size == 0` would fix it. For example:

```
Initial:
  size = 0;
  front = 0;
  back = 0;

Push:
  front = 0;
  back = if size == 0 { 0 } else { size - 1 } = 0;
  size = size + 1 = 1;
  prev = back = 0;
  back = (back + 1) % size = 0;

Query:
  agg = buffer[back] = buffer[0]

Pop:
  front = (front + 1) % 1 = 0;
  size = size - 1 = 0;
```

Reply:
My concern is that when `size == 1`, `front == 0`, `back == 1`, is that in `query()`, I think `tracing_indices` will be `[0]` after the building loop. And then in the popping loop, we'll look at `buffer[0].val`, which is incorrect. The final statement will look at `buffer[1].val`, which is correct, but at that point, `agg` has already been built up from an unitialized value. Or am I mistaken?

It might be worthwhile to add this as a simple test - push into an empty window, query the result, check that it is correct.

Reply:
Hmm, I am wondering if it is ever possible for `back == 1` when `size == 1`. In either case, I added some additional tests.

Reply:
Yes, you're correct. I was mistaken in my reasoning about `back`; I kept thinking "a small value of `back` plus one mod the size is just `back + 1`", but of course, that's not true when `back == 0` and `size == 1`.